File: src/2015/day02/python/solution.py
import math


def format_input(input: str) -> list[list[int]]:
    dimensions = [
        [int(dim) for dim in line.split("x")] for line in input.strip().splitlines()
    ]
    return dimensions


def calculate_paper(dimensions: list[int]) -> int:

    sides = (
        dimensions[0] * dimensions[1],
        dimensions[1] * dimensions[2],
        dimensions[2] * dimensions[0],
    )
    min_side = min(sides)

    total_square_feet = sum(map(lambda x: x * 2, sides))
    return int(min_side + total_square_feet)


def calculate_ribbon(dimensions: list[int]) -> int:
    min_two = sorted(dimensions)[0:2]
    face_ribbon = sum(map(lambda d: d * 2, min_two))
    bow_ribbon = math.prod(dimensions)
    return face_ribbon + bow_ribbon


def part1(input: str) -> int:
    paper = sum([calculate_paper(dimensions) for dimensions in format_input(input)])
    return paper


def part2(input: str) -> int:
    paper = sum([calculate_ribbon(dimensions) for dimensions in format_input(input)])
    return paper

File: src/2015/day02/typescript/solution.ts
type Dimensions = number[];

function formatInput(input: string): Dimensions[] {
  return input
    .trim()
    .split('\n')
    .map((line) => line.trim().split('x').map(Number));
}

function calculatePaper(dimensions: Dimensions): number {
  const [l, w, h] = dimensions;

  const sides = [l * w, w * h, h * l];
  const minSide = Math.min(...sides);

  const totalSurfaceArea = 2 * (sides[0] + sides[1] + sides[2]);

  return minSide + totalSurfaceArea;
}

function calculateRibbon(dimensions: Dimensions): number {
  const sortedDims = dimensions.toSorted((a, b) => a - b);
  const faceRibbon = 2 * (sortedDims[0] + sortedDims[1]);
  const bowRibbon = dimensions.reduce((prod, dimension) => prod * dimension, 1);
  return faceRibbon + bowRibbon;
}

export function part1(input: string): number | string | undefined {
  const totalPaper = formatInput(input)
    .map((row) => calculatePaper(row))
    .reduce((sum, num) => sum + num, 0);
  return totalPaper;
}

export function part2(input: string): number | string | undefined {
  const totalRibbon = formatInput(input)
    .map((row) => calculateRibbon(row))
    .reduce((sum, num) => sum + num, 0);
  return totalRibbon;
}
